### PR TITLE
[DNM] rgw: Don't increment bi list entry count on error to not distort error code

### PR DESCRIPTION
**Pull request by haklein**
http://tracker.ceph.com/issues/21205

### CODE REVIEW

**Comment by cbodley** [comment]
it looks like some jewel backports were applied in the wrong order, resulting in this `+ 1; /* one extra entry for identifying truncation */` piece being duplicated (it's present in the line that calls `list_plain_entries()` and above in the initialization of `int32_t max = ...`)

looking at the history of master, the two relevant commits were applied in this order:
- 47f422a4e0382d53023af6f651433011606b8625 cls/rgw: bi_list() fix is_truncated returned param
- 51a4405a1ef59671cbd51c530a84333359d3f3dd rgw: fix osd crashes when execute "radosgw-admin bi list --max-entries=1" command

on jewel, the backports were applied in the opposite order:
- 8de67afed415875394ad0405046ead6206ffa636 rgw: fix osd crashes when execute "radosgw-admin bi list --max-entries=1" command
- e197ec0312cb07d576c0bf045b803454702ccc81 cls/rgw: bi_list() fix is_truncated returned param

the correct resolution should restore this to what we have on master:
```c++
#define MAX_BI_LIST_ENTRIES 1000
  int32_t max = (op.max < MAX_BI_LIST_ENTRIES ? op.max : MAX_BI_LIST_ENTRIES) + 1; /* one extra entry for identifying truncation */
  string start_key = op.marker;
  int ret = list_plain_entries(hctx, op.name, op.marker, max, &op_ret.entries);
  if (ret < 0) {
    CLS_LOG(0, "ERROR: %s(): list_plain_entries retured ret=%d", __func__, ret);
    return ret;
  }
  int count = ret;
```

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -2520,12 +2520,12 @@ static int rgw_bi_list_op(cls_method_context_t hctx, bufferlist *in, bufferlist
...
   int32_t max = (op.max < MAX_BI_LIST_ENTRIES ? op.max : MAX_BI_LIST_ENTRIES) + 1; /* one extra entry for identifying truncation */
   string start_key = op.marker;
-  int ret = list_plain_entries(hctx, op.name, op.marker, max, &op_ret.entries) + 1; /* one extra entry for identifying truncation */
+  int ret = list_plain_entries(hctx, op.name, op.marker, max, &op_ret.entries);
   if (ret < 0) {
     CLS_LOG(0, "ERROR: %s(): list_plain_entries retured ret=%d", __func__, ret);
     return ret;
   }
-  int count = ret;
+  int count = ret + 1; /* one extra entry for identifying truncation */;
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw.cc]: this line should not be changed. the `+ 1` is already accounted for above with `int32_t max = ...`

**Comment by smithfarm** [comment]
Thanks for the fix, @haklein - can you make the change requested by @cbodley?

**Comment by smithfarm** [comment]
This passed an rgw suite at http://tracker.ceph.com/issues/20613#note-43

**Comment by smithfarm** [comment]
Replacing with https://github.com/ceph/ceph/pull/18207

**Pull request closed without merging**